### PR TITLE
feat(rpc-types-engine): no-std support

### PR DESCRIPTION
**Pull request by refcell**
### Description

> [!NOTE]
>
> Blocked by https://github.com/alloy-rs/alloy/pull/1252

Adds `no_std` for the `alloy-rpc-types-engine` crate.

### CODE REVIEW

File: crates/rpc-types-engine/src/lib.rs [hunk 1]
```diff
@@ -5,6 +5,10 @@
 )]
 #![cfg_attr(not(test), warn(unused_crate_dependencies))]
 #![cfg_attr(docsrs, feature(doc_cfg, doc_auto_cfg))]
+#![cfg_attr(not(feature = "jwt"), no_std)]
```
Comment (mattsse) [inline crates/rpc-types-engine/src/lib.rs:8]: this should use std feature to disable no_std

File: crates/rpc-types-engine/src/lib.rs [hunk 2]
```diff
@@ -5,6 +5,10 @@
...
+
+#[cfg(not(feature = "jwt"))]
```
Comment (mattsse) [inline crates/rpc-types-engine/src/lib.rs:10]: same

File: crates/rpc-types-engine/Cargo.toml [hunk 3]
```diff
@@ -43,7 +43,7 @@ rand = { workspace = true, optional = true }
 
 [features]
 default = ["jwt"]
```
Comment (mattsse) [inline crates/rpc-types-engine/Cargo.toml:45]: this needs a std feature

**Review by mattsse — CHANGES_REQUESTED** [review]
let's also add this crate to

https://github.com/alloy-rs/op-alloy/blob/bd669da1ea0239411b317d81e30a0477aabb0808/scripts/check_no_std.sh#L4-L5

so we check in ci

File: crates/rpc-types-engine/src/payload.rs [hunk 4]
```diff
@@ -1,7 +1,7 @@
 //! Payload types.
 
-#[cfg(not(feature = "jwt"))]
-use alloc::vec::Vec;
+#[cfg(not(feature = "std"))]
+use alloc::{string::{ToString, String}, vec::Vec};
```
Comment (mattsse) [inline crates/rpc-types-engine/src/payload.rs]: let's use extern crate alloc so we dont need to feature gate these

File: crates/rpc-types-engine/src/payload.rs [hunk 5]
```diff
@@ -771,72 +771,75 @@ impl<'de> Deserialize<'de> for ExecutionPayload {
...
     InvalidVersionedHashes,
     /// Encountered decoding error.
-    #[error(transparent)]
-    Decode(#[from] alloy_rlp::Error),
+    #[display("{_0}")]
+    Decode(alloy_rlp::Error),
 }
 
+#[cfg(feature = "std")]
+impl std::error::Error for PayloadError {}
```
Comment (mattsse) [inline crates/rpc-types-engine/src/payload.rs:844]: this needs to impl fn source and return the rlp error

File: crates/rpc-types-engine/src/payload.rs [hunk 6]
```diff
@@ -771,72 +771,75 @@ impl<'de> Deserialize<'de> for ExecutionPayload {
...
     },
     /// Expected blob versioned hashes do not match the given transactions.
-    #[error("expected blob versioned hashes do not match the given transactions")]
+    #[display("expected blob versioned hashes do not match the given transactions")]
     InvalidVersionedHashes,
     /// Encountered decoding error.
-    #[error(transparent)]
-    Decode(#[from] alloy_rlp::Error),
+    #[display("{_0}")]
+    Decode(alloy_rlp::Error),
```
Comment (mattsse) [inline crates/rpc-types-engine/src/payload.rs:840]: this is now missing the from impl

File: crates/rpc-types-engine/src/forkchoice.rs [hunk 7]
```diff
@@ -34,22 +34,25 @@ pub struct ForkchoiceState {
 ///
 /// These are considered hard RPC errors and are _not_ returned as [PayloadStatus] or
 /// [PayloadStatusEnum::Invalid].
-#[derive(Clone, Copy, Debug, PartialEq, Eq, thiserror::Error)]
+#[derive(Clone, Copy, Debug, PartialEq, Eq, derive_more::Display)]
```
Comment (mattsse) [inline crates/rpc-types-engine/src/forkchoice.rs:37]: once we bump msrv to 1.81 we can also use derive_more error and remove the feature entirely

https://docs.rs/derive_more/latest/derive_more/derive.Error.html

**Review by mattsse — CHANGES_REQUESTED** [review]
nits

**Comment by refcell** [comment]
Replaced with https://github.com/alloy-rs/alloy/pull/1268

**Pull request closed without merging**